pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferHandlerImpl.java

```diff
@@ -116,24 +124,26 @@ public TransactionBufferHandlerImpl(PulsarClient pulsarClient,
 
     private CompletableFuture<TxnID> endTxn(long requestId, String topic, ByteBuf cmd, CompletableFuture<TxnID> cb) {
         OpRequestSend op = OpRequestSend.create(requestId, topic, cmd, cb);
-        pendingRequests.put(requestId, op);
-        cmd.retain();
         try {
             cache.get(topic).whenComplete((clientCnx, throwable) -> {
                 if (throwable == null) {
-                    try {
+                    if (clientCnx.ctx().channel().isActive()) {
                         clientCnx.registerTransactionBufferHandler(TransactionBufferHandlerImpl.this);
+                        synchronized (TransactionBufferHandlerImpl.class) {
+                            pendingRequests.put(requestId, op);
+                            cmd.retain();
+                        }
                         clientCnx.ctx().writeAndFlush(cmd, clientCnx.ctx().voidPromise());
-                    } catch (Exception e) {
+                    } else {
                         cache.invalidate(topic);
-                        cb.completeExceptionally(e);
-                        pendingRequests.remove(requestId);
+                        cb.completeExceptionally(
+                                new PulsarClientException.LookupException(topic + " endTxn channel is not active"));
                         op.recycle();
                     }
                 } else {
+                    log.error("endTxn error topic: [{}]", topic, throwable);
                     cache.invalidate(topic);
                     cb.completeExceptionally(throwable);
-                    pendingRequests.remove(requestId);
```

Comment:
Why are you removing this line?

Reply:
if run here, the request don't put in `pendingRequests`. so remove this line

Reply:
makes sense, thanks